d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -833,4 +854,59 @@ private ImmutablePair<String, String> getZnodePathAndData(String cluster) {
   public boolean isWarmingUp() {
     return _isWarmingUp;
   }
+
+  private Map<Integer, PartitionData> validatePartitionData(Map<Integer, PartitionData> partitionData) {
+    if (_maxWeight == null || _actOnWeightBreach == ActOnWeightBreach.IGNORE) {
+      return partitionData;
+    }
+
+    Map<Integer, PartitionData> res = new HashMap<>(partitionData); // modifiable copy in case the input is unmodifiable
+    for (Map.Entry<Integer, PartitionData> entry : res.entrySet()) {
+      BigDecimal weight = BigDecimal.valueOf(entry.getValue().getWeight());
+      // check max weight
+      if (weight.compareTo(_maxWeight) > 0) {
+        switch (_actOnWeightBreach) {
+          case WARN:
+            _log.warn("", getMaxWeightBreachException(weight, entry.getKey()));
+            break;
+          case THROW:
+            throw getMaxWeightBreachException(weight, entry.getKey());
+          case RECTIFY:
+            entry.setValue(new PartitionData(_maxWeight.intValue()));
```

Comment:
Interesting. This value will be overridded by following code, right ? Let's say the weight is `101.123`, the current logic result will be `101.12`

Reply:
yeah, if the value breaches decimal places, it will be rounded. If it breaches the max, it will be capped. I updated the log files in the container PR with sample behaviors. Please take a look, thanks.

Reply:
I think @brycezhongqing might be saying that if the value both breaches decimal places **and** breaches the max, will only the rounding take effect but not the capping?

Because we do `entry.setValue` but we don't update the `weight` variable

Reply:
Yes. @ZoabKapoor 
So @bohhyang If a weight value both breaches both `max value and decimal places`, there will be a potential risk for current logic.

Reply:
oh, that's right. Thanks for catching!